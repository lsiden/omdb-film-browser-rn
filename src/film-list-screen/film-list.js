import React from "react"
import PropTypes from "prop-types"
import { connect } from "react-redux"
import { FlatList } from "react-native"

import { fetchQueryResultPage } from "actions/fetch"
import Header from "./header"
import Footer from "./footer"
import { Screens } from "constants"
import FilmListItem from "./film-list-item"

export const filmList = ({
  navigation,
  films,
  totalResults,
  dispatchFetchPage
}) => {
  const navigateToDetailsScreen = navigation.navigate.bind(
    null,
    Screens.FilmDetails
  )

  // eslint-disable-next-line react/prop-types
  const renderItem = ({ item }) => (
    <FilmListItem
      filmSummary={item}
      navigateToDetailsScreen={navigateToDetailsScreen}
    />
  )
  return (
    <FlatList
      data={films}
      renderItem={renderItem}
      keyExtractor={item => item.imdbID}
      ListHeaderComponent={<Header totalResults={totalResults} />}
      ListFooterComponent={<Footer />}
      onEndReached={() => dispatchFetchPage()}
      initialNumToRender={8}
      onEndReachedThreshold={1}
    />
  )
}

filmList.propTypes = {
  navigation: PropTypes.object.isRequired,
  films: PropTypes.arrayOf(PropTypes.object),
  totalResults: PropTypes.number,
  dispatchFetchPage: PropTypes.func.isRequired
}
filmList.defaultProps = {
  isFetching: false
}

export default connect(
  state => ({
    films: state.films || [],
    totalResults: state.totalResults || 0
  }),
  dispatch => ({
    dispatchFetchPage: () => dispatch(fetchQueryResultPage())
  })
)(filmList)
