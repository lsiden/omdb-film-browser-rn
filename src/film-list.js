import React from "react"
import PropTypes from "prop-types"
import { connect } from "react-redux"
import { Text, View, ActivityIndicator, StyleSheet } from "react-native"
import { List, ListItem } from "react-native-elements"

import { filmListStyles } from "./styles"
import { fetchFilmDetails } from "./actions/fetch"
import { viewFilmSummary, updateIsFetching } from "./actions"
import Pager from "components/pager"
import LoadingIndicator from "./components/loading-indicator"

const renderMessage = msg => <Text style={filmListStyles.msgStyle}>{msg}</Text>

export const filmList = ({
  films,
  dispatchViewDetail,
  isFetching,
  totalResults,
}) => {
  if (isFetching) {
    return <LoadingIndicator />
  } else if (films.length > 0) {
    return (
      <View>
        {renderMessage(`Found ${totalResults} results.`)}
        <List>
          {films.map(filmSummary => (
            <ListItem
              key={filmSummary.imdbID}
              title={filmSummary.Title}
              subtitle={filmSummary.Year}
              onPress={() => dispatchViewDetail(filmSummary)}
            />
          ))}
        </List>
        <Pager />
      </View>
    )
  } else {
    return renderMessage("There are no films that match your query yet.")
  }
}

filmList.propTypes = {
  films: PropTypes.arrayOf(PropTypes.object),
  dispatchViewDetail: PropTypes.func.isRequired,
  isFetching: PropTypes.bool,
  totalResults: PropTypes.number,
}
filmList.defaultProps = {
  films: [],
  isFetching: false,
}

export default connect(
  state => ({
    films: state.films,
    isFetching: state.isFetching,
    totalResults: state.totalResults,
  }),
  dispatch => ({
    dispatchViewDetail: filmSummary => {
      dispatch(updateIsFetching(true))
      dispatch(viewFilmSummary(filmSummary))
      dispatch(fetchFilmDetails(filmSummary.imdbID))
    },
  })
)(filmList)
