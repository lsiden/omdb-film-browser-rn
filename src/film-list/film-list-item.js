/* NOTE
 * This is a lot of boilerplate for such a small copmonent.
 * I broke FilmListItem out of FilmList in the hope of being able to test that
 * dispatchViewDetails() property gets called when a ListItem is pressed.
 * Unfortunately, react-native-elements renders ListItem in a way
 * that does not allow me to simulate a press in my test
 * or to find and invokes the onPress() property directly.
 */
import React from "react"
import PropTypes from "prop-types"
import { connect } from "react-redux"
import { ListItem } from "react-native-elements"

import { viewFilmDetails } from "actions"
import { fetchFilmDetails } from "actions/fetch"

export const filmListItem = ({ filmSummary, dispatchViewDetails }) => (
  <ListItem
    title={filmSummary.Title}
    subtitle={filmSummary.Year}
    onPress={() => dispatchViewDetails(filmSummary)}
  />
)

filmListItem.propTypes = {
  filmSummary: PropTypes.object.isRequired,
  dispatchViewDetails: PropTypes.func.isRequired,
}

// Export this to enable mocking.
export default connect(null, dispatch => ({
  dispatchViewDetails: filmSummary => {
    dispatch(viewFilmDetails())
    dispatch(fetchFilmDetails(filmSummary.imdbID))
  },
}))(filmListItem)
