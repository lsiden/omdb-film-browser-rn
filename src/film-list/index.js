import React from "react"
import PropTypes from "prop-types"
import { connect } from "react-redux"
import { Text, View } from "react-native"
import { List } from "react-native-elements"

import { appStyles, loadingIndicatorStyle } from "styles"
import Pager from "components/pager"
import LoadingIndicator from "components/loading-indicator"
import FilmListItem from "./film-list-item"

const renderMessage = msg => <Text style={appStyles.msgStyle}>{msg}</Text>
const loadingWrapperStyle = {
  height: "100%",
  justifyContent: "center",
}
export const filmList = ({ films, isFetching, totalResults }) => {
  if (isFetching) {
    return (
      <View style={loadingWrapperStyle}>
        <LoadingIndicator style={loadingIndicatorStyle} />
      </View>
    )
  } else if (films.length > 0) {
    return (
      <View>
        {renderMessage(`Found ${totalResults} results.`)}
        <List>
          {films.map(filmSummary => (
            <FilmListItem key={filmSummary.imdbID} filmSummary={filmSummary} />
          ))}
        </List>
        <Pager />
      </View>
    )
  } else {
    return renderMessage("There are no films that match your query yet.")
  }
}

filmList.propTypes = {
  films: PropTypes.arrayOf(PropTypes.object),
  isFetching: PropTypes.bool,
  totalResults: PropTypes.number,
}
filmList.defaultProps = {
  films: [],
  isFetching: false,
}

export default connect(state => ({
  films: state.films,
  isFetching: state.isFetching,
  totalResults: state.totalResults,
}))(filmList)
