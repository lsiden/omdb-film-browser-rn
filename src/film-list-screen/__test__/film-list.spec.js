import React from "react"
import renderer from "react-native-test-utils"

import { filmList as FilmList } from "../film-list"

const films = require("__fixture__/films.json").Search

jest.mock("actions/fetch", () => ({
  fetchFilmDetails: () => "fetchFilmDetails"
}))
jest.mock("../film-list-item", () => "FilmListItem")
jest.mock("../header", () => "Header")
jest.mock("../footer", () => "Footer")

// This does not help
jest.mock("abort-controller")
jest.mock("actions/fetch")

const defaultProps = () => {
  return {
    films,
    pageNum: 1,
    isFetching: false,
    dispatchFetchPage: () => {},
    dispatchFetchFilmDetails: () => {},
    dispatchResetFilmDetails: () => {},
    navigation: {
      navigate: () => {}
    }
  }
}

const createWrapper = (props = {}) =>
  renderer(<FilmList {...{ ...defaultProps(), ...props }} />)

it("renders list of titles", () => {
  const wrapper = createWrapper()
  expect(wrapper.toJSON()).toMatchSnapshot()
})
