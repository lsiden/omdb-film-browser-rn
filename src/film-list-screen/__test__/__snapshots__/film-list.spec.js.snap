// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders list of titles 1`] = `
<RCTScrollView
  ListFooterComponent={<Footer />}
  ListHeaderComponent={
    <Header
      totalResults={undefined}
    />
  }
  data={
    Array [
      Object {
        "Poster": "https://images-na.ssl-images-amazon.com/images/M/MV5BODAzMDgxMDc1MF5BMl5BanBnXkFtZTgwMTI0OTAzMjE@._V1_SX300.jpg",
        "Title": "The Hobbit: The Battle of the Five Armies",
        "Type": "movie",
        "Year": "2014",
        "imdbID": "tt2310332",
      },
      Object {
        "Poster": "https://images-na.ssl-images-amazon.com/images/M/MV5BMDg2NzQwOGMtMGRkNC00YjAwLTg4NjgtZWQwYzljZmM1YzA4XkEyXkFqcGdeQXVyNTIzOTk5ODM@._V1_SX300.jpg",
        "Title": "Battle Los Angeles",
        "Type": "movie",
        "Year": "2011",
        "imdbID": "tt1217613",
      },
      Object {
        "Poster": "https://images-na.ssl-images-amazon.com/images/M/MV5BMWI1YmQ0ZTctZWM2YS00YWE3LWE0NWUtNGE5YTFiNDlhOTY4L2ltYWdlXkEyXkFqcGdeQXVyMzM4MjM0Nzg@._V1_SX300.jpg",
        "Title": "Battle Royale",
        "Type": "movie",
        "Year": "2000",
        "imdbID": "tt0266308",
      },
      Object {
        "Poster": "https://images-na.ssl-images-amazon.com/images/M/MV5BMGZjZDE2NjUtMzVkNy00N2UyLTg5MjYtYzY1ZmM1YWY3ZTZmXkEyXkFqcGdeQXVyNjk1Njg5NTA@._V1_SX300.jpg",
        "Title": "Night at the Museum: Battle of the Smithsonian",
        "Type": "movie",
        "Year": "2009",
        "imdbID": "tt1078912",
      },
      Object {
        "Poster": "https://images-na.ssl-images-amazon.com/images/M/MV5BYzliNzI5NzgtNDZmNy00NGU4LTg2OTQtMDEzZTRiNDU3ZTI1XkEyXkFqcGdeQXVyMDUyOTUyNQ@@._V1_SX300.jpg",
        "Title": "The Battle of Algiers",
        "Type": "movie",
        "Year": "1966",
        "imdbID": "tt0058946",
      },
      Object {
        "Poster": "https://images-na.ssl-images-amazon.com/images/M/MV5BZTljYmU2NTMtODhhNC00NjlhLWJhZTUtNDllODYyYWM4ZjA5XkEyXkFqcGdeQXVyNjM0ODk5NDY@._V1_SX300.jpg",
        "Title": "Battle of the Sexes",
        "Type": "movie",
        "Year": "2017",
        "imdbID": "tt4622512",
      },
      Object {
        "Poster": "https://images-na.ssl-images-amazon.com/images/M/MV5BNWRlYjU3ODYtNzAzZS00YWE5LTkxZTEtYjM1YzIwZTlhMWQ2L2ltYWdlL2ltYWdlXkEyXkFqcGdeQXVyMTQxNzMzNDI@._V1_SX300.jpg",
        "Title": "Battle for the Planet of the Apes",
        "Type": "movie",
        "Year": "1973",
        "imdbID": "tt0069768",
      },
      Object {
        "Poster": "https://images-na.ssl-images-amazon.com/images/M/MV5BMjE2MzUwODUxMl5BMl5BanBnXkFtZTgwMTI0NDkzMjE@._V1_SX300.jpg",
        "Title": "Dragon Ball Z: Battle of Gods",
        "Type": "movie",
        "Year": "2013",
        "imdbID": "tt2263944",
      },
      Object {
        "Poster": "http://ia.media-imdb.com/images/M/MV5BMjE2MzUwODUxMl5BMl5BanBnXkFtZTgwMTI0NDkzMjE@._V1_SX300.jpg",
        "Title": "Dragon Ball Z: Battle of Gods",
        "Type": "movie",
        "Year": "2013",
        "imdbID": "tt2536562",
      },
      Object {
        "Poster": "https://images-na.ssl-images-amazon.com/images/M/MV5BYjY1NmY1MzUtOTgwYy00Mzk0LThkNzgtZGRhNWU4NmFmYzMyL2ltYWdlXkEyXkFqcGdeQXVyNjQzNDI3NzY@._V1_SX300.jpg",
        "Title": "Battle of Britain",
        "Type": "movie",
        "Year": "1969",
        "imdbID": "tt0064072",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={8}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onContentSizeChange={[Function]}
  onEndReached={[Function]}
  onEndReachedThreshold={1}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <Header
        totalResults={undefined}
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <FilmListItem
        filmSummary={
          Object {
            "Poster": "https://images-na.ssl-images-amazon.com/images/M/MV5BODAzMDgxMDc1MF5BMl5BanBnXkFtZTgwMTI0OTAzMjE@._V1_SX300.jpg",
            "Title": "The Hobbit: The Battle of the Five Armies",
            "Type": "movie",
            "Year": "2014",
            "imdbID": "tt2310332",
          }
        }
        navigateToDetailsScreen={[Function]}
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <FilmListItem
        filmSummary={
          Object {
            "Poster": "https://images-na.ssl-images-amazon.com/images/M/MV5BMDg2NzQwOGMtMGRkNC00YjAwLTg4NjgtZWQwYzljZmM1YzA4XkEyXkFqcGdeQXVyNTIzOTk5ODM@._V1_SX300.jpg",
            "Title": "Battle Los Angeles",
            "Type": "movie",
            "Year": "2011",
            "imdbID": "tt1217613",
          }
        }
        navigateToDetailsScreen={[Function]}
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <FilmListItem
        filmSummary={
          Object {
            "Poster": "https://images-na.ssl-images-amazon.com/images/M/MV5BMWI1YmQ0ZTctZWM2YS00YWE3LWE0NWUtNGE5YTFiNDlhOTY4L2ltYWdlXkEyXkFqcGdeQXVyMzM4MjM0Nzg@._V1_SX300.jpg",
            "Title": "Battle Royale",
            "Type": "movie",
            "Year": "2000",
            "imdbID": "tt0266308",
          }
        }
        navigateToDetailsScreen={[Function]}
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <FilmListItem
        filmSummary={
          Object {
            "Poster": "https://images-na.ssl-images-amazon.com/images/M/MV5BMGZjZDE2NjUtMzVkNy00N2UyLTg5MjYtYzY1ZmM1YWY3ZTZmXkEyXkFqcGdeQXVyNjk1Njg5NTA@._V1_SX300.jpg",
            "Title": "Night at the Museum: Battle of the Smithsonian",
            "Type": "movie",
            "Year": "2009",
            "imdbID": "tt1078912",
          }
        }
        navigateToDetailsScreen={[Function]}
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <FilmListItem
        filmSummary={
          Object {
            "Poster": "https://images-na.ssl-images-amazon.com/images/M/MV5BYzliNzI5NzgtNDZmNy00NGU4LTg2OTQtMDEzZTRiNDU3ZTI1XkEyXkFqcGdeQXVyMDUyOTUyNQ@@._V1_SX300.jpg",
            "Title": "The Battle of Algiers",
            "Type": "movie",
            "Year": "1966",
            "imdbID": "tt0058946",
          }
        }
        navigateToDetailsScreen={[Function]}
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <FilmListItem
        filmSummary={
          Object {
            "Poster": "https://images-na.ssl-images-amazon.com/images/M/MV5BZTljYmU2NTMtODhhNC00NjlhLWJhZTUtNDllODYyYWM4ZjA5XkEyXkFqcGdeQXVyNjM0ODk5NDY@._V1_SX300.jpg",
            "Title": "Battle of the Sexes",
            "Type": "movie",
            "Year": "2017",
            "imdbID": "tt4622512",
          }
        }
        navigateToDetailsScreen={[Function]}
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <FilmListItem
        filmSummary={
          Object {
            "Poster": "https://images-na.ssl-images-amazon.com/images/M/MV5BNWRlYjU3ODYtNzAzZS00YWE5LTkxZTEtYjM1YzIwZTlhMWQ2L2ltYWdlL2ltYWdlXkEyXkFqcGdeQXVyMTQxNzMzNDI@._V1_SX300.jpg",
            "Title": "Battle for the Planet of the Apes",
            "Type": "movie",
            "Year": "1973",
            "imdbID": "tt0069768",
          }
        }
        navigateToDetailsScreen={[Function]}
      />
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <FilmListItem
        filmSummary={
          Object {
            "Poster": "https://images-na.ssl-images-amazon.com/images/M/MV5BMjE2MzUwODUxMl5BMl5BanBnXkFtZTgwMTI0NDkzMjE@._V1_SX300.jpg",
            "Title": "Dragon Ball Z: Battle of Gods",
            "Type": "movie",
            "Year": "2013",
            "imdbID": "tt2263944",
          }
        }
        navigateToDetailsScreen={[Function]}
      />
    </View>
    <View
      style={
        Object {
          "height": 0,
        }
      }
    />
    <View
      onLayout={[Function]}
      style={null}
    >
      <Footer />
    </View>
  </View>
</RCTScrollView>
`;
